### PR TITLE
Throw an error if the current Julia version (`Base.VERSION`) is incompatible with the `[compat]` entry for `julia` in the `Project.toml` file

### PR DESCRIPTION
**Pull request by Krastanov**
Pkg now raises an error if the `[compat]` entry for `julia` is not met for a package.

This change was made due to new developers getting confused why their julia 1.8 does not work with a dev-ed package that has a julia compat set to 1.9.

There is no escape hatch, i.e. this behavior can not be disabled.

### CODE REVIEW

**Comment by Krastanov** [comment]
`Base.get_bool_env` is not present in julia 1.9... I assume that is fine given that `Base.get_bool_env` is used elsewhere in `Pkg`.

**Comment by DilumAluthge** [comment]
Even if the user has set `JULIA_PKG_CHECK_JULIA_COMPAT`, we should still print a warning, in my opinion.

**Comment by Krastanov** [comment]
The TODO that I deleted used to say "either raise an error or be completely silent", so I did not want to overstep and followed those guidelines. Happy to change it though. What would be the best way for me to ask all stakeholders about their opinion? Should I tag someone here?

**Comment by Krastanov** [comment]
Pinging @KristofferC  as he is the one who wrote the "error or be quiet" TODO in https://github.com/JuliaLang/Pkg.jl/pull/2179

**Comment by Krastanov** [comment]
@KristofferC , @fredrikekre , just a friendly bump on this relatively minor change. Could you give me a rough idea whether this is something to be pursued and polished for merging or to be abandoned if it is contrary to how Pkg is supposed to work?

**Comment by Krastanov** [comment]
bump?

**Comment by DilumAluthge** [comment]
Bump @KristofferC @fredrikekre

**Review by IanButterworth — CHANGES_REQUESTED** [review]
This seems like a good idea. It needs a test though. Perhaps create a test package that has an upper bound on an old julia version?

**Comment by IanButterworth** [comment]
Nice. The authors of these test packages have me laughing 😂

**Comment by IanButterworth** [comment]
Great. Can you also add a note to the NEWS.md

**Comment by DilumAluthge** [comment]
Can we bikeshed the name of the environment variable?

**Comment by IanButterworth** [comment]
Maybe it should be the reverse?

`JULIA_PKG_JULIA_COMPAT_IGNORE=1` defaulting to off.

**Comment by Krastanov** [comment]
I will wait a day before making changes to the variable name and add NEWS.md entry. Let me know if there are other suggestions.

**Comment by Krastanov** [comment]
uhm... where is the NEWS.md? Maybe I am a bit sleep deprived, but I can not seem to find it

**Comment by Krastanov** [comment]
I see a changelog file, but should I add this as a 1.10 or 1.11 entry?

**Comment by IanButterworth** [comment]
Changelog, my bad. And 1.11

**Comment by IanButterworth** [comment]
To keep track of this I'm assigning @DilumAluthge here to help finish this up, given he raised discussion around some of this on slack

**Comment by Krastanov** [comment]
I renamed the env variable (and made it logical reverse, so it is set to 1 to ignore), added a changelog, and removed the `=$set_value` in the error message.

**Comment by DilumAluthge** [comment]
Based on a conversation in Slack, what if the three values for the environment variable are:
1. `error` (the default value)
2. `warn`
3. `ignore`

**Comment by DilumAluthge** [comment]
Also, @IanButterworth this only applies to the `Project.toml` file of the active project, right?

File: src/Operations.jl
Comment (DilumAluthge) [inline src/Operations.jl:248]:
```suggestion
    if julia_compat !== nothing && !(VERSION in julia_compat)
        # If the environment variable is not set, or if it is set to "error", we throw an error.
        # If the environment variable is set to "warn", we print a warning, and we continue.
        # If the environment variable is set to "ignore", we do not print any warning, and we continue.
        #
        # If the environment variable is set to any other value, we print a message saying that the
        # user has set an invalid value for the environment variable, and then we proceed as if
        # the environment variable was set to "error", i.e. we throw an error.
        env_var_name = "JULIA_PKG_CHECK_JULIA_COMPAT"
        env_var_value = strip(get(ENV, env_var_name, "error"))
        msg = "Julia version requirement for package $(err_rep(pkg)) not satisfied; you can override and ignore this check by setting `ENV[\"$(env_var_name)\"]`"
        if env_var_value == "warn"
            @warn msg
        elseif env_var_value == "ignore"
        else
            if env_var_value != "error"
                @warn "Invalid value for $(env_var_name): $(env_var_value)"
            end
            pkgerror(msg)
        end
    end
```

**Review by DilumAluthge — CHANGES_REQUESTED** [review]

File: src/Operations.jl
Comment (DilumAluthge) [reply src/Operations.jl]: I was under the impression that this is only for the active project's `Project.toml`, but I might be mistaken.

@IanButterworth Can you answer this?

**Comment by DilumAluthge** [comment]
I'll leave the `needs changelog` label on for now, since we'll have to update the changelog once we've finalized the plan for the environment variable and its values.

File: CHANGELOG.md
Comment (DilumAluthge) [inline CHANGELOG.md:1]: @IanButterworth Should we have an empty "Pkg v1.10 Release Notes" section, just for consistency?

File: CHANGELOG.md
Comment (DilumAluthge) [inline CHANGELOG.md:5]:
```suggestion
  The old behavior of disregarding Julia `compat` bounds can be restored by setting the `JULIA_PKG_JULIA_COMPAT_IGNORE` environmental variable.
```

**Comment by DilumAluthge** [comment]
@Krastanov and @IanButterworth What do you guys think of `error`/`warn`/`silent` versus `error`/`warn`/`ignore`?

**Comment by Krastanov** [comment]
I am happy to keep the env variable as it is or to implement error/warn/silent|ignore. I would prefer to be given a "benevolent dictator" decision from above as I do not feel I have a "feel" for the "culture" of this codebase.

File: src/Operations.jl
Comment (Krastanov) [reply src/Operations.jl:248]: let me know what is decided on this and I will be happy to do the finishing touches

**Comment by Krastanov** [comment]
I completely removed the env variable as suggested on slack. Happy to undo that if someone enforces a different preference

File: CHANGELOG.md
Comment (DilumAluthge) [inline CHANGELOG.md:4]:
```suggestion
- Pkg now obeys `[compat]` bounds for `julia` and raises an error if the version of the running Julia binary is incompatible with the bounds in `Project.toml`.
```

**Comment by Krastanov** [comment]
I think all change requests are fulfilled. Let me know if anything is missing.

**Comment by DilumAluthge** [comment]
I think this is good to go, but I'll do one last review this weekend to make sure.

**Comment by Krastanov** [comment]
bump

File: test/api.jl
Comment (DilumAluthge) [inline test/api.jl:395]:
```suggestion
@testset "`[compat]` entries for `julia`" begin
```

**Review by DilumAluthge — APPROVED** [review]
One small suggested change, but otherwise LGTM.

**Review by DilumAluthge — APPROVED** [review]

**Comment by DilumAluthge** [comment]
@IanButterworth I'll leave it to you to do the final review, and to click the merge button.

**Review by IanButterworth — APPROVED** [review]
This has been discussed on slack and been open here for a while and I haven't seen any disagreement, so seems worth giving a go for 1.11. Thanks @Krastanov

**Comment by DilumAluthge** [comment]
Thanks @Krastanov!

**Comment by DilumAluthge** [comment]
And thanks @IanButterworth for sheparding this thing through!

**Pull request merged**
